src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,187 @@
+        {
+            await EnsureBufferSpaceAsync(8, isAsync, ct).ConfigureAwait(false);
+            BinaryPrimitives.WriteInt64LittleEndian(GetWritableSpan(8), v);
+            _outBytesUsed += 8;
+        }
+
+        /// <summary>
+        /// Writes unsigned long value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public Task WriteUnsignedLongAsync(ulong v, bool isAsync, CancellationToken ct) => WriteLongAsync((long)v, isAsync, ct);
+
+        /// <summary>
+        /// Writes double value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public async Task WriteDoubleAsync(double v, bool isAsync, CancellationToken ct)
+        {
+            if (double.IsInfinity(v) || double.IsNaN(v))
+            {
+                throw new ArgumentOutOfRangeException(nameof(v), "Double value is out of range.");
+            }
+            await EnsureBufferSpaceAsync(8, isAsync, ct).ConfigureAwait(false);
+            BinaryPrimitives.WriteInt64LittleEndian(GetWritableSpan(8), BitConverter.DoubleToInt64Bits(v));
+            _outBytesUsed += 8;
+        }
+
+        /// <summary>
+        /// Writes partial long value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="length">Length to fill</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public async Task WritePartialLongAsync(long v, int length, bool isAsync, CancellationToken ct)
+        {
+            if (length > 8)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length), "Length specified is longer than the size of a long");
+            }
+            else if (length < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length), "Length should not be negative");
+            }
+
+            await EnsureBufferSpaceAsync(length, isAsync, ct).ConfigureAwait(false);
+            Memory<byte> memory = GetWritableMemory(length);
+            for (int i = 0; i < length; i++)
+            {
+                memory.Span[i] = (byte)((v >> (i * 8)) & 0xFF);
+            }
+            _outBytesUsed += length;
+        }
+        #endregion
+
+        #region Private helpers
+        private async Task EnsureBufferSpaceAsync(int requiredBytes, bool isAsync, CancellationToken ct)
+        {
+            if ((_outBytesUsed + requiredBytes) > _outBuff.Length)
+            {
+                if (isAsync)
+                {
+                    await _writeStream.WriteAsync(_outBuff.AsMemory(0, _outBytesUsed), ct).ConfigureAwait(false);
```

Comment:
For any writes to the stream, you dont have to split data into _outbytes etc.
You can dump an arbitrarily large array of bytes to it, and it will chunk the bytes and wrap it into packets and send it out